pkg/skaffold/schema/v1alpha2/profiles.go

```diff
@@ -18,47 +18,47 @@ package v1alpha2
 
 import (
 	"fmt"
+	"reflect"
 
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
-	yaml "gopkg.in/yaml.v2"
 )
 
 // ApplyProfiles returns configuration modified by the application
 // of a list of profiles.
 func (c *SkaffoldConfig) ApplyProfiles(profiles []string) error {
-	var err error
-
 	byName := profilesByName(c.Profiles)
 	for _, name := range profiles {
 		profile, present := byName[name]
 		if !present {
 			return fmt.Errorf("couldn't find profile %s", name)
 		}
 
-		err = applyProfile(c, profile)
-		if err != nil {
-			return errors.Wrapf(err, "applying profile %s", name)
-		}
+		applyProfile(c, profile)
 	}
-
-	c.Profiles = nil
 	if err := c.setDefaultValues(); err != nil {
 		return errors.Wrap(err, "applying default values")
 	}
 
 	return nil
 }
 
-func applyProfile(config *SkaffoldConfig, profile Profile) error {
-	logrus.Infof("Applying profile: %s", profile.Name)
+func applyProfile(config *SkaffoldConfig, profile Profile) {
+	logrus.Infof("applying profile: %s", profile.Name)
 
-	buf, err := yaml.Marshal(profile)
-	if err != nil {
-		return err
+	// this intentionally removes the Profiles field from the returned config
+	*config = SkaffoldConfig{
+		APIVersion: config.APIVersion,
+		Kind:       config.Kind,
+		Build: BuildConfig{
+			Artifacts: overlayProfileField(config.Build.Artifacts, profile.Build.Artifacts).([]*Artifact),
```

Comment:
this is an interesting - if I understand correctly you are avoiding to have to do a recursion/iteration on the profile structure by manually picking fields, right? The problem with this is that when we change the config structure, we will have to touch this code too because it introduces a dependency on the structure and decouples profile from ... if we'd add a new field, e.g.  config.Build.Foo, tests wouldn't fail, right, so we have to remember to add it here too?

Reply:
The issue here is that most of the values we're overlaying here are themselves structs (not primitive values), so we can't know which ones to treat as "values" and which ones to treat as "holders of values" (and recurse over them). This solution implicitly chooses the first non-nil value from the profile and uses it as the new value in the final config.

```golang
if fieldValue != nil && !reflect.ValueOf(fieldValue).IsNil() {
    // use this value and return
}
```

For example, the `BuildConfig` struct (top level in the `SkaffoldConfig`) contains the `TagPolicy` struct, which itself is composed of several pointers to structs:

```
type TagPolicy struct {
	GitTagger         *GitTagger         `yaml:"gitCommit"`
	ShaTagger         *ShaTagger         `yaml:"sha256"`
	EnvTemplateTagger *EnvTemplateTagger `yaml:"envTemplate"`
	DateTimeTagger    *DateTimeTagger    `yaml:"dateTime"`
}
```
`GitTagger` is technically of type `struct`, but we treat it as an actual value. Since we say it's invalid to have multiple values set here, this code just picks the first non-nil one and uses it. This was handled by yaml marshalling before because any value that wasn't set would be set to `nil` by marshalling the yaml, but if we add `yaml:omitempty` tags to the config this will break.

If we add something to the top level fields in the `SkaffoldConfig`, then it's true that tests could still pass and we could potentially miss a silent failure. I'm not sure of a good way to address this apart from either 1) reworking the implementation of the config to not use pointers as a way to implement "one-of" values (see https://github.com/GoogleContainerTools/skaffold/pull/388), or 2) being more rigorous with our testing when we make changes to the config. WDYT?

Reply:
Thanks for the detailed explanation! 
- Can we somehow mark these top level elements as "one-of" and recognize that in the `switch v.Kind()`? it would be cool :) (#388 does it on the "option" level not on the "struct" level, maybe we could sync on this @dlorenc ? )
- I would split the `overlayProfileField` to `overlayOneOfField` and `overlayField` methods because the fact that the struct fields currently in the profile are "one-of" types has nothing to do inherently with the profile, it is confusing.

Reply:
> Can we somehow mark these top level elements as "one-of" and recognize that in the switch v.Kind()? it would be cool :) (#388 does it on the "option" level not on the "struct" level, maybe we could sync on this @dlorenc ? )

Marking them is doable with YAML tags, it should be relatively easy to add in here. I'll chat with Nick.